go/rpcclientlib/obscuro_client.go

```diff
@@ -19,39 +29,228 @@ const (
 }
 
-func NewClient(address string) Client {
+func NewClient(address string) *RPCClient {
+	// todo: this is a convenience for testnet but needs to replaced by a parameter and/or retrieved from the target host
+	enclPubECDSA, err := crypto.DecompressPubkey(common.Hex2Bytes(enclavePublicKeyHex))
+	if err != nil {
+		panic(err)
+	}
+	enclavePublicKey := ecies.ImportECDSAPublic(enclPubECDSA)
+
 	rpcClient, err := rpc.Dial(http + address)
 	if err != nil {
 		log.Panic("could not create RPC client on %s. Cause: %s", http+address, err)
 	}
 
-	return &clientImpl{
-		rpcClient: rpcClient,
+	return &RPCClient{
+		rpcClient:        rpcClient,
+		enclavePublicKey: enclavePublicKey,
+	}
+}
+
+// Call handles JSON rpc requests - if the method is sensitive it will encrypt the args before sending the request and
+//		then decrypts the response before returning
+// The result must be a pointer so that package json can unmarshal into it. You
+// 	can also pass nil, in which case the result is ignored.
+func (c *RPCClient) Call(result interface{}, method string, args ...interface{}) error {
+	if !isSensitive(method) {
+		// for non-sensitive methods or when viewing keys are disabled we just delegate directly to the geth RPC client
+		return c.rpcClient.Call(&result, method, args...)
+	}
+
+	// we setup a generic rawResult to receive the response (then we can decrypt it as necessary into the requested result type)
+	var err error
+	var rawResult interface{}
+	if result == nil {
+		// we set result receiver to nil if caller set nil since no return needs to be set
+		rawResult = nil
+	}
+
+	if len(args) == 0 {
+		// no encryption necessary
+		err = c.rpcClient.Call(&rawResult, method)
+	} else {
+		// encode the params into a json blob and encrypt them
+		var encryptedParams []byte
+		encryptedParams, err = c.encryptArgs(args...)
+		if err != nil {
+			return fmt.Errorf("failed to encrypt args for rpc call - %w", err)
+		}
+		err = c.rpcClient.Call(&rawResult, method, encryptedParams)
+	}
+	if err != nil {
+		return fmt.Errorf("rpc call failed - %w", err)
+	}
+
+	// if caller not interested in response, we're done
+	if result == nil {
+		return nil
+	}
+
+	// todo: remove this non-decryption part when we make viewing key encryption mandatory across all tests
+	// extract result from the data as-is, in case we can't decrypt/process it below
+	decrypted, ok := rawResult.([]byte) // if viewing key was nil we try and extract result from the data as-is
+	if !ok {
+		decStr, ok := rawResult.(string)
+		if ok {
+			decrypted = []byte(decStr)
+		}
+	}
+	if c.viewingKey != nil {
+		// method is sensitive, so we decrypt it before unmarshalling the result
+		decrypted, err = c.decryptResponse(rawResult)
+		if err != nil {
+			return fmt.Errorf("failed to decrypt args for rpc call - %w", err)
+		}
+	}
+
+	// process the decrypted result to get the desired type and set it on the result pointer
+	err = c.prepareResult(decrypted, result)
```

Comment:
@BedrockSquirrel Can this get a better name? `setResult` is better than `prepareResult`, I think.